File: src/main/java/com/dansoftware/libraryapp/gui/info/InformationView.java
package com.dansoftware.libraryapp.gui.info;

import com.dansoftware.libraryapp.gui.context.Context;
import com.dansoftware.libraryapp.gui.info.dependency.DependencyViewerActivity;
import com.dansoftware.libraryapp.gui.util.ImprovedFXMLLoader;
import com.dansoftware.libraryapp.locale.I18N;
import com.dansoftware.libraryapp.locale.LanguageTranslator;
import com.dlsc.workbenchfx.SimpleHeaderView;
import com.dlsc.workbenchfx.view.controls.ToolbarItem;
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon;
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIconView;
import javafx.event.ActionEvent;
import javafx.event.Event;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.control.Tooltip;
import javafx.scene.effect.DropShadow;
import javafx.scene.input.Clipboard;
import javafx.scene.input.ClipboardContent;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.net.URL;
import java.util.Locale;
import java.util.ResourceBundle;

import static com.dansoftware.libraryapp.gui.info.ApplicationInfoCopyKt.getApplicationInfoCopy;

/**
 * An InformationView is the actual GUI element that shows the
 * information of the application.
 *
 * @author Daniel Gyorffy
 */
public class InformationView extends SimpleHeaderView<Node>
        implements Initializable {

    private final Context context;

    @FXML
    private Label versionLabel;

    @FXML
    private Label javaVMLabel;

    @FXML
    private Label javaVendorLabel;

    @FXML
    private Label javaVersionLabel;

    @FXML
    private Label javaHomeLabel;

    @FXML
    private Label javaFXVersionLabel;

    @FXML
    private Label langLabel;

    @FXML
    private Label langTranslatorLabel;

    @FXML
    private Button gitHubBtn;

    @FXML
    private Button showDepsBtn;

    @FXML
    private TextField logsLocationField;


    public InformationView(@NotNull Context context) {
        super("LibraryApp Info", new MaterialDesignIconView(MaterialDesignIcon.INFORMATION));
        this.context = context;
        this.setEffect(new DropShadow());
        this.createToolbarControls();
        this.loadContent();
    }

    @FXML
    private void showUsedDependencies(ActionEvent event) {
        new DependencyViewerActivity(this.context).show();
    }

    private void createToolbarControls() {
        var copyItem = new ToolbarItem(
                StringUtils.EMPTY,
                new MaterialDesignIconView(MaterialDesignIcon.CONTENT_COPY),
                this::onCopy
        );
        copyItem.setTooltip(new Tooltip(I18N.getInfoViewValues().getString("info.copy")));
        this.getToolbarControlsRight().add(copyItem);
    }

    private void loadContent() {
        var fxmlLoader = new ImprovedFXMLLoader(this, getClass().getResource("View.fxml"), I18N.getInfoViewValues());
        this.setContent(fxmlLoader.load());
    }

    private void onCopy(Event event) {
        ClipboardContent content = new ClipboardContent();
        content.putString(getApplicationInfoCopy());
        Clipboard clipboard = Clipboard.getSystemClipboard();
        clipboard.setContent(content);
    }

    private void setData() {
        versionLabel.setText(System.getProperty("libraryapp.version"));
        javaVMLabel.setText(System.getProperty("java.vm.name"));
        javaVendorLabel.setText(System.getProperty("java.vendor"));
        javaHomeLabel.setText(System.getProperty("java.home"));
        javaVersionLabel.setText(System.getProperty("java.version"));
        javaFXVersionLabel.setText(System.getProperty("javafx.version"));
        logsLocationField.setText(System.getProperty("log.file.path.full"));
        langLabel.setText(Locale.getDefault().getDisplayLanguage());
        LanguageTranslator translator = I18N.getLanguagePack().getTranslator();
        if (translator != null) langTranslatorLabel.setText(translator.getDisplayName(Locale.getDefault()));
        else langTranslatorLabel.setText("?");
    }

    private void setDefaults() {
        gitHubBtn.setGraphic(new MaterialDesignIconView(MaterialDesignIcon.GITHUB_BOX));
        showDepsBtn.setGraphic(new MaterialDesignIconView(MaterialDesignIcon.CODE_BRACES));
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        setDefaults();
        setData();
    }
}

File: src/main/java/com/dansoftware/libraryapp/gui/info/InformationView.kt
package com.dansoftware.libraryapp.gui.info

import com.dansoftware.libraryapp.gui.context.Context
import com.dansoftware.libraryapp.gui.context.TitledOverlayBox
import com.dansoftware.libraryapp.gui.info.dependency.DependencyViewerActivity
import com.dansoftware.libraryapp.gui.util.HighlightableLabel
import com.dansoftware.libraryapp.i18n.I18N
import com.dansoftware.libraryapp.util.SystemBrowser
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIconView
import javafx.geometry.Insets
import javafx.scene.Group
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.input.Clipboard
import javafx.scene.input.ClipboardContent
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox
import jfxtras.styles.jmetro.JMetroStyleClass
import org.apache.commons.lang3.StringUtils
import java.util.*

class InformationViewOverlay(context: Context) :
    TitledOverlayBox(
        I18N.getInfoViewValues().getString("info.view.title"),
        MaterialDesignIconView(MaterialDesignIcon.INFORMATION),
        InformationView(context),
        resizableH = true,
        resizableV = false,
        Button().also {
            it.contentDisplay = ContentDisplay.GRAPHIC_ONLY
            it.graphic = MaterialDesignIconView(MaterialDesignIcon.CONTENT_COPY)
            it.tooltip = Tooltip(I18N.getInfoViewValues().getString("info.copy"))
            it.setOnAction {
                ClipboardContent().also { clipboardContent ->
                    clipboardContent.putString(getApplicationInfoCopy())
                    Clipboard.getSystemClipboard().setContent(clipboardContent)
                }
            }
            it.padding = Insets(0.0)
        }
    )

class InformationView(val context: Context) : VBox(5.0) {

    companion object {
        const val GITHUB_REPO_URL = "https://github.com/Dansoftowner/LibraryApp2020"
    }

    init {
        styleClass.add("information-view")
        styleClass.add(JMetroStyleClass.BACKGROUND)
        buildProgramInfo()
        separator()
        buildJavaInfo()
        separator()
        buildLogInfo()
        separator()
        buildBottom()
    }

    private fun separator() = children.add(Separator())

    private fun buildProgramInfo() {
        children.add(KeyValuePair("info.version", System.getProperty("libraryapp.version")))
        children.add(KeyValuePair("info.developer", "Györffy Dániel"))
        children.add(KeyValuePair("info.lang", Locale.getDefault().displayLanguage))
        children.add(
            KeyValuePair(
                "info.lang.translator",
                I18N.getLanguagePack().translator?.getDisplayName(Locale.getDefault())
            )
        )
    }

    private fun buildJavaInfo() {
        children.add(KeyValuePair("java.home", System.getProperty("java.home")))
        children.add(KeyValuePair("java.vm", System.getProperty("java.vm.name").let { vmName ->
            System.getProperty("java.vm.vendor").let {
                if (StringUtils.isBlank(it).not()) {
                    "$vmName ${I18N.getInfoViewValues().getString("java.vm.by")} $it"
                } else null
            } ?: vmName
        }))
        children.add(KeyValuePair("java.version", System.getProperty("java.version")))
        children.add(KeyValuePair("javafx.version", System.getProperty("javafx.version")))
    }

    private fun buildLogInfo() {
        children.add(KeyValuePair("info.logs.loc", System.getProperty("log.file.path.full")))
    }

    private fun buildBottom() = StackPane(Group(HBox(10.0).also {
        it.children.add(buildGithubButton())
        it.children.add(buildDependencyButton())
    })).also { children.add(it) }

    private fun buildDependencyButton() =
        Button(
            I18N.getInfoViewValues().getString("info.show.dependencies"),
            MaterialDesignIconView(MaterialDesignIcon.CODE_BRACES)
        ).also {
            it.setOnAction {
                DependencyViewerActivity(this.context).show()
            }
        }

    private fun buildGithubButton() =
        Button(
            I18N.getInfoViewValues().getString("info.show.github"),
            MaterialDesignIconView(MaterialDesignIcon.GITHUB_BOX)
        ).also {
            it.setOnAction {
                SystemBrowser.browse(GITHUB_REPO_URL)
            }
        }


    private class KeyValuePair(i18n: String, value: Node) : HBox() {

        constructor(i18n: String, value: String?) : this(i18n, HighlightableLabel(value).also {
            setHgrow(it, Priority.ALWAYS)
        })

        init {
            children.add(Label(I18N.getInfoViewValues().getString(i18n)))
            children.add(Label(":").also {
                it.padding = Insets(0.0, 5.0, 0.0, 0.0)
            })
            children.add(value)
        }

    }
}